File: packages/pdp-extractor-agent/src/llm/providers/azure-openai.ts
import { createImageOcrPrompt, createKeywordClassificationPromptParts } from "../prompt";
import {
  chatCompletionsImageOcrResponseFormat,
  chatCompletionsKeywordClassificationResponseFormat
} from "../schemas";
import type {
  ImageTextExtractionRequest,
  ImageTextExtractionResponse,
  KeywordClassificationRequest,
  KeywordClassificationResponse,
  KeywordClassifier,
  LlmProviderConfig
} from "../types";
import type { AiTokenUsage } from "../../types";
import {
  downloadImageAsDataUrl,
  extractJsonObjectText,
  fetchWithTimeout,
  isUnsupportedStructuredOutputError,
  mergeTokenUsages,
  numberField,
  parseImageOcrPayloadText,
  resolveImageInputs,
  responseErrorSuffix,
  sumOptional,
  temperatureBody
} from "./shared";

export { temperatureBody } from "./shared";

type ChatCompletionsPayload = {
  choices?: Array<{ message?: { content?: string } }>;
  usage?: unknown;
};

const CHAT_COMPLETIONS_TIMEOUT_MS = 300_000;

/** Azure API adapter using deployment-scoped chat completions. */
export class AzureApiKeywordClassifier implements KeywordClassifier {
  constructor(protected readonly config: LlmProviderConfig) {}

  /** Builds the chat completions URL for a deployment. Overridden by gateway adapters with a different path/auth scheme. */
  protected chatCompletionsUrl(deployment: string): string {
    const apiVersion = this.config.apiVersion ?? "2025-04-01-preview";
    const endpoint = this.config.endpoint?.replace(/\/$/, "");
    return `${endpoint}/openai/deployments/${deployment}/chat/completions?api-version=${apiVersion}`;
  }

  /** Auth headers for deployment-scoped calls. Azure uses the api-key header; gateways may override with a Bearer token. */
  protected authHeaders(): Record<string, string> {
    return { "api-key": this.config.apiKey ?? "" };
  }

  async classifyKeywords(request: KeywordClassificationRequest): Promise<KeywordClassificationResponse> {
    const deployment = this.config.deployments?.reasoning ?? this.config.deployment;
    if (!this.config.apiKey || !this.config.endpoint || !deployment) {
      throw new Error("API key, endpoint, and a reasoning deployment/model id are required.");
    }

    const url = this.chatCompletionsUrl(deployment);
    const prompt = createKeywordClassificationPromptParts(request);
    const payload = await requestChatCompletionsJson(
      url,
      this.authHeaders(),
      {
        messages: [
          { role: "system", content: prompt.system },
          { role: "user", content: prompt.user }
        ],
        response_format: chatCompletionsKeywordClassificationResponseFormat,
        ...temperatureBody(this.config.temperature)
      },
      "Azure keyword classification"
    );
    const rawText = payload.choices?.[0]?.message?.content ?? "";
    const jsonText = extractJsonObjectText(rawText);
    const result = jsonText ? JSON.parse(jsonText) as KeywordClassificationResponse : { keywords: [], summary: "No parseable JSON returned.", rawText };
    return {
      ...result,
      usage: tokenUsageFromChatCompletions(payload.usage)
    };
  }

  async extractImageTexts(request: ImageTextExtractionRequest): Promise<ImageTextExtractionResponse> {
    const deployment = this.config.deployments?.ocr ?? this.config.deployment;
    if (!this.config.apiKey || !this.config.endpoint || !deployment) {
      throw new Error("API key, endpoint, and an OCR deployment/model id are required for image OCR.");
    }

    const directImages = resolveImageInputs(request);

    try {
      const direct = await this.requestImageOcr(request, directImages, deployment);
      const missingInputs = imageInputsWithoutExtractedText(directImages, direct.images);

      if (missingInputs.length === 0) {
        return direct;
      }

      const fallback = await this.extractImageTextsWithDownloadedImages(
        { ...request, imageUrls: missingInputs.map((input) => input.displayUrl), imageInputs: missingInputs },
        deployment,
        new Error(`${missingInputs.length} remote image OCR result(s) returned no readable text.`)
      );

      return fallback.images.length > 0
        ? mergeImageTextExtractionResponses(direct, fallback)
        : direct;
    } catch (directError) {
      const fallback = await this.extractImageTextsWithDownloadedImages(request, deployment, directError);

      if (fallback.images.length > 0) {
        return fallback;
      }

      throw directError;
    }
  }

  private async requestImageOcr(
    request: ImageTextExtractionRequest,
    images: Array<{ displayUrl: string; inputUrl: string }>,
    deployment: string
  ): Promise<ImageTextExtractionResponse> {
    const url = this.chatCompletionsUrl(deployment);
    const payload = await requestChatCompletionsJson(
      url,
      this.authHeaders(),
      {
        messages: [
          {
            role: "user",
            content: createImageOcrContent({
              ...request,
              imageUrls: images.map((image) => image.displayUrl)
            }, images)
          }
        ],
        response_format: chatCompletionsImageOcrResponseFormat,
        ...temperatureBody(this.config.temperature)
      },
      "Azure image OCR"
    );
    return {
      ...parseImageOcrPayloadText(payload.choices?.[0]?.message?.content ?? "", images.map((image) => image.displayUrl)),
      usage: tokenUsageFromChatCompletions(payload.usage)
    };
  }

  private async extractImageTextsWithDownloadedImages(
    request: ImageTextExtractionRequest,
    deployment: string,
    directError: unknown
  ): Promise<ImageTextExtractionResponse> {
    const images: ImageTextExtractionResponse["images"] = [];
    const errors: string[] = [];
    const usages: AiTokenUsage[] = [];

    for (const input of resolveImageInputs(request)) {
      try {
        const dataUrl = input.inputUrl.startsWith("data:") ? input.inputUrl : await downloadImageAsDataUrl(input.inputUrl);
        const extracted = await this.requestImageOcr(request, [{ displayUrl: input.displayUrl, inputUrl: dataUrl }], deployment);
        images.push(...extracted.images);
        if (extracted.usage) {
          usages.push(extracted.usage);
        }
      } catch (error) {
        errors.push(error instanceof Error ? error.message : String(error));
      }
    }

    return {
      images,
      rawText: [
        directError instanceof Error ? directError.message : String(directError),
        ...errors
      ].join("\n"),
      usage: mergeTokenUsages(usages)
    };
  }
}

function createImageOcrContent(
  request: ImageTextExtractionRequest,
  images: Array<{ displayUrl: string; inputUrl: string }>
) {
  return [
    {
      type: "text",
      text: createImageOcrPrompt({
        ...request,
        imageUrls: images.map((image) => image.displayUrl)
      })
    },
    ...images.flatMap((image, index) => [
      {
        type: "text",
        text: `Image ${index + 1}: ${image.displayUrl}`
      },
      {
        type: "image_url",
        image_url: {
          url: image.inputUrl,
          detail: "high"
        }
      }
    ])
  ];
}

function imageInputsWithoutExtractedText(
  requestedInputs: Array<{ displayUrl: string; inputUrl: string }>,
  extractedImages: ImageTextExtractionResponse["images"]
): Array<{ displayUrl: string; inputUrl: string }> {
  const extractedUrls = new Set(extractedImages.map((image) => image.imageUrl));
  return requestedInputs.filter((input) => !extractedUrls.has(input.displayUrl));
}

function mergeImageTextExtractionResponses(
  primary: ImageTextExtractionResponse,
  fallback: ImageTextExtractionResponse
): ImageTextExtractionResponse {
  const seen = new Set<string>();
  const images = [...primary.images, ...fallback.images].filter((image) => {
    const key = image.imageUrl;
    if (seen.has(key)) {
      return false;
    }
    seen.add(key);
    return image.text.trim().length > 0;
  });

  return {
    images,
    rawText: [primary.rawText, fallback.rawText].filter(Boolean).join("\n"),
    usage: mergeTokenUsages([primary.usage, fallback.usage].filter((usage): usage is AiTokenUsage => Boolean(usage)))
  };
}

async function requestChatCompletionsJson(
  url: string,
  authHeaders: Record<string, string>,
  body: Record<string, unknown>,
  failureLabel: string
): Promise<ChatCompletionsPayload> {
  let currentBody = body;
  let response = await postChatCompletions(url, authHeaders, currentBody, failureLabel);

  if (!response.ok) {
    const suffix = await responseErrorSuffix(response);

    if (currentBody.response_format !== undefined && isUnsupportedStructuredOutputError(suffix)) {
      const { response_format: _responseFormat, ...retryBody } = currentBody;
      currentBody = retryBody;
      response = await postChatCompletions(url, authHeaders, currentBody, failureLabel);
    } else if (currentBody.temperature !== undefined && isUnsupportedTemperatureError(suffix)) {
      const { temperature: _temperature, ...retryBody } = currentBody;
      currentBody = retryBody;
      response = await postChatCompletions(url, authHeaders, currentBody, failureLabel);
    } else {
      throw new Error(`${failureLabel} failed: ${response.status}${suffix}`);
    }
  }

  if (!response.ok) {
    const suffix = await responseErrorSuffix(response);

    if (currentBody.temperature !== undefined && isUnsupportedTemperatureError(suffix)) {
      const { temperature: _temperature, ...retryBody } = currentBody;
      const retryResponse = await postChatCompletions(url, authHeaders, retryBody, failureLabel);

      if (retryResponse.ok) {
        return retryResponse.json() as Promise<ChatCompletionsPayload>;
      }

      throw new Error(`${failureLabel} failed: ${retryResponse.status}${await responseErrorSuffix(retryResponse)}`);
    }

    throw new Error(`${failureLabel} failed: ${response.status}${suffix}`);
  }

  return response.json() as Promise<ChatCompletionsPayload>;
}

function postChatCompletions(
  url: string,
  authHeaders: Record<string, string>,
  body: Record<string, unknown>,
  failureLabel: string
): Promise<Response> {
  return fetchWithTimeout(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...authHeaders
    },
    body: JSON.stringify(body)
  }, CHAT_COMPLETIONS_TIMEOUT_MS, failureLabel);
}

function isUnsupportedTemperatureError(message: string): boolean {
  return /unsupported value[^]*temperature|temperature[^]*(?:unsupported|only the default)/i.test(message);
}

function tokenUsageFromChatCompletions(value: unknown): AiTokenUsage | undefined {
  if (!value || typeof value !== "object") {
    return undefined;
  }
  const usage = value as Record<string, unknown>;
  const inputTokens = numberField(usage.prompt_tokens) ?? numberField(usage.input_tokens);
  const outputTokens = numberField(usage.completion_tokens) ?? numberField(usage.output_tokens);
  const totalTokens = numberField(usage.total_tokens) ?? sumOptional(inputTokens, outputTokens);
  return inputTokens !== undefined || outputTokens !== undefined || totalTokens !== undefined
    ? { inputTokens, outputTokens, totalTokens }
    : undefined;
}

File: packages/pdp-extractor-agent/src/pdp_extractor_agent/providers/azure_openai.py
"""Azure deployment-scoped chat-completions provider compatibility adapter."""

from __future__ import annotations

import json
import math
from collections.abc import Awaitable, Callable, Mapping
from typing import Any, cast

import httpx
from neo_js_compat import js_json_bytes

from .._json_types import as_list, as_mapping
from ..ocr.layout import parse_image_ocr_payload_text
from ..ocr.prompt import create_image_ocr_prompt, create_keyword_classification_prompt_parts
from ..schemas import (
    CHAT_COMPLETIONS_IMAGE_OCR_RESPONSE_FORMAT,
    CHAT_COMPLETIONS_KEYWORD_CLASSIFICATION_RESPONSE_FORMAT,
)
from .shared import data_url
from .transport import (
    DEFAULT_MODEL_TIMEOUT_SECONDS,
    is_unsupported_structured_output_error,
    model_timeout_seconds,
    resolve_image_inputs,
    response_json_object,
    temperature_body,
    timeout_label_seconds,
)

ImageFetcher = Callable[[str], Awaitable[tuple[str, bytes]]]


class AzureOpenAIProvider:
    """Deployment-addressed Azure chat/OCR adapter with legacy retries."""

    def __init__(
        self,
        *,
        api_key: object,
        endpoint: object,
        deployment: object | None = None,
        deployments: Mapping[str, object] | None = None,
        api_version: object | None = None,
        temperature: object = None,
        transport: object = None,
        image_fetcher: object = None,
        timeout_seconds: object = DEFAULT_MODEL_TIMEOUT_SECONDS,
    ) -> None:
        # Do not validate or normalize these injected values at construction
        # time. The factory historically passed application adapters through
        # untouched; request construction is the first meaningful boundary.
        self.api_key = api_key
        self.endpoint = endpoint
        self.deployment = deployment
        self.deployments = dict(deployments or {})
        self.api_version = api_version
        self.temperature = temperature
        self.transport = transport
        self.image_fetcher = image_fetcher
        self.timeout_seconds = model_timeout_seconds(timeout_seconds)

    def chat_completions_url(self, deployment: object) -> str:
        version = self.api_version if self.api_version is not None else "2025-04-01-preview"
        # Base Azure preserves the legacy direct interpolation contract.  AI
        # Studio intentionally has separate trimming/component-encoding rules.
        endpoint = _text(self.endpoint).removesuffix("/")
        return f"{endpoint}/openai/deployments/{deployment}/chat/completions?api-version={version}"

    def auth_headers(self) -> dict[str, str]:
        return {"api-key": _text(self.api_key)}

    async def classify_keywords(self, request: Mapping[str, Any]) -> dict[str, Any]:
        deployment = _nullish(self.deployments.get("reasoning"), self.deployment)
        self._require(deployment, "reasoning")
        prompt = create_keyword_classification_prompt_parts(request)
        payload = await self._request_chat_completions_json(
            self.chat_completions_url(deployment),
            {
                "messages": [
                    {"role": "system", "content": prompt["system"]},
                    {"role": "user", "content": prompt["user"]},
                ],
                "response_format": CHAT_COMPLETIONS_KEYWORD_CLASSIFICATION_RESPONSE_FORMAT,
                **temperature_body(self.temperature),
            },
            "Azure keyword classification",
        )
        raw_text = _chat_content(payload)
        result = _parse_keyword_payload(raw_text)
        usage = _token_usage(as_mapping(payload.get("usage")))
        if usage is not None:
            result["usage"] = usage
        return result

    async def classifyKeywords(self, request: Mapping[str, Any]) -> dict[str, Any]:
        return await self.classify_keywords(request)

    async def extract_image_text(self, request: Mapping[str, Any]) -> dict[str, Any]:
        deployment = _nullish(self.deployments.get("ocr"), self.deployment)
        self._require(deployment, "OCR", image=True)
        inputs = resolve_image_inputs(request)
        try:
            direct = await self._request_image_ocr(request, inputs, deployment)
            missing = [item for item in inputs if item["displayUrl"] not in _extracted_urls(direct)]
            if not missing:
                return direct
            fallback = await self._downloaded_fallback(
                {**dict(request), "imageUrls": [item["displayUrl"] for item in missing], "imageInputs": missing},
                deployment,
                RuntimeError(f"{len(missing)} remote image OCR result(s) returned no readable text."),
            )
            return _merge_image_responses(direct, fallback) if _has_image_text(fallback) else direct
        except Exception as direct_error:
            fallback = await self._downloaded_fallback(request, deployment, direct_error)
            if _has_image_text(fallback):
                return fallback
            raise

    async def extractImageTexts(self, request: Mapping[str, Any]) -> dict[str, Any]:
        return await self.extract_image_text(request)

    async def _request_image_ocr(
        self, request: Mapping[str, Any], inputs: list[dict[str, str]], deployment: object
    ) -> dict[str, Any]:
        content: list[dict[str, Any]] = [
            {
                "type": "text",
                "text": create_image_ocr_prompt({**dict(request), "imageUrls": [item["displayUrl"] for item in inputs]}),
            }
        ]
        for index, item in enumerate(inputs, start=1):
            content.extend(
                [
                    {"type": "text", "text": f"Image {index}: {item['displayUrl']}"},
                    {"type": "image_url", "image_url": {"url": item["inputUrl"], "detail": "high"}},
                ]
            )
        payload = await self._request_chat_completions_json(
            self.chat_completions_url(deployment),
            {
                "messages": [{"role": "user", "content": content}],
                "response_format": CHAT_COMPLETIONS_IMAGE_OCR_RESPONSE_FORMAT,
                **temperature_body(self.temperature),
            },
            "Azure image OCR",
        )
        raw_text = _chat_content(payload)
        result = parse_image_ocr_payload_text(raw_text, [item["displayUrl"] for item in inputs])
        usage = _token_usage(as_mapping(payload.get("usage")))
        if usage is not None:
            result["usage"] = usage
        return result

    async def _downloaded_fallback(
        self, request: Mapping[str, Any], deployment: object, direct_error: BaseException
    ) -> dict[str, Any]:
        images: list[dict[str, Any]] = []
        errors: list[str] = []
        usages: list[Mapping[str, Any]] = []
        for input_ in resolve_image_inputs(request):
            try:
                input_url = input_["inputUrl"]
                data_input = input_url if input_url.startswith("data:") else await self._download_image_as_data_url(input_url)
                extracted = await self._request_image_ocr(
                    request, [{"displayUrl": input_["displayUrl"], "inputUrl": data_input}], deployment
                )
                images.extend(_image_rows(extracted))
                usage = as_mapping(extracted.get("usage"))
                if usage is not None:
                    usages.append(usage)
            except Exception as error:
                errors.append(str(error))
        result: dict[str, Any] = {
            "images": images,
            "rawText": "\n".join([str(direct_error), *errors]),
        }
        merged_usage = _merge_usages(usages)
        if merged_usage is not None:
            result["usage"] = merged_usage
        return result

    async def _request_chat_completions_json(
        self, url: str, body: Mapping[str, Any], label: str
    ) -> Mapping[str, Any]:
        current = dict(body)
        response = await self._post(url, current, label)
        if not response.is_success:
            suffix = _response_error_suffix(response)
            if "response_format" in current and is_unsupported_structured_output_error(suffix):
                current.pop("response_format", None)
                response = await self._post(url, current, label)
            elif "temperature" in current and _unsupported_temperature(suffix):
                current.pop("temperature", None)
                response = await self._post(url, current, label)
            else:
                raise RuntimeError(f"{label} failed: {response.status_code}{suffix}")
        if not response.is_success:
            suffix = _response_error_suffix(response)
            if "temperature" in current and _unsupported_temperature(suffix):
                retry = dict(current)
                retry.pop("temperature", None)
                response = await self._post(url, retry, label)
                if response.is_success:
                    return response_json_object(response, label)
                raise RuntimeError(f"{label} failed: {response.status_code}{_response_error_suffix(response)}")
            raise RuntimeError(f"{label} failed: {response.status_code}{suffix}")
        return response_json_object(response, label)

    async def _post(self, url: str, body: Mapping[str, Any], label: str) -> httpx.Response:
        headers = {"Content-Type": "application/json", **self.auth_headers()}
        async with httpx.AsyncClient(
            transport=cast(httpx.AsyncBaseTransport | None, self.transport),
            timeout=self.timeout_seconds,
            follow_redirects=True,
        ) as client:
            try:
                return await client.post(url, headers=headers, content=js_json_bytes(body))
            except httpx.TimeoutException as error:
                raise RuntimeError(f"{label} timed out after {timeout_label_seconds(self.timeout_seconds)}s.") from error

    async def _download_image_as_data_url(self, image_url: str) -> str:
        if callable(self.image_fetcher):
            fetched = cast(ImageFetcher, self.image_fetcher)(image_url)
            mime_type, content = await fetched
            return data_url(mime_type, content)
        async with httpx.AsyncClient(
            transport=cast(httpx.AsyncBaseTransport | None, self.transport), timeout=60.0, follow_redirects=True
        ) as client:
            try:
                response = await client.get(
                    image_url,
                    headers={
                        "Accept": "image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8",
                        "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0 Safari/537.36",
                    },
                )
            except httpx.TimeoutException as error:
                raise RuntimeError("Image download for OCR timed out after 60s.") from error
        if not response.is_success:
            raise RuntimeError(f"Image download failed: {response.status_code}{_response_error_suffix(response)}")
        mime_type = response.headers.get("content-type", "image/jpeg").split(";", 1)[0].strip() or "image/jpeg"
        if not mime_type.startswith("image/"):
            raise RuntimeError(f"Image download returned non-image content-type: {mime_type}")
        if len(response.content) > 10 * 1024 * 1024:
            raise RuntimeError(f"Image is too large for OCR fallback: {len(response.content)} bytes")
        return data_url(mime_type, response.content)

    def _require(self, deployment: object, role: str, *, image: bool = False) -> None:
        if not self.api_key or not self.endpoint or not deployment:
            suffix = " are required for image OCR." if image else " are required."
            article = "an" if role == "OCR" else "a"
            raise ValueError(f"API key, endpoint, and {article} {role} deployment/model id{suffix}")


def _parse_keyword_payload(raw_text: str) -> dict[str, Any]:
    json_text = _extract_json_object_text(raw_text)
    if json_text is None:
        return {"keywords": [], "summary": "No parseable JSON returned.", "rawText": raw_text}
    parsed = json.loads(json_text)
    return dict(as_mapping(parsed) or {})


def _extract_json_object_text(raw_text: str) -> str | None:
    text = raw_text.strip()
    if text.startswith("{") and text.endswith("}"):
        return text
    fenced = re_search_fenced_json(text)
    if fenced is not None and fenced.startswith("{"):
        return fenced
    match = __import__("re").search(r"\{[\s\S]*\}", text)
    return match.group(0) if match is not None else None


def re_search_fenced_json(text: str) -> str | None:
    match = __import__("re").search(r"```(?:json)?\s*([\s\S]*?)\s*```", text, __import__("re").IGNORECASE)
    return match.group(1).strip() if match is not None else None


def _chat_content(payload: Mapping[str, Any]) -> str:
    choices = as_list(payload.get("choices")) or []
    first = as_mapping(choices[0]) if choices else None
    message = as_mapping(first.get("message")) if first is not None else None
    content = message.get("content") if message is not None else ""
    return content if isinstance(content, str) else ""


def _token_usage(value: Mapping[str, Any] | None) -> dict[str, int | float] | None:
    if value is None:
        return None
    input_tokens = _number_field(value.get("prompt_tokens"))
    if input_tokens is None:
        input_tokens = _number_field(value.get("input_tokens"))
    output_tokens = _number_field(value.get("completion_tokens"))
    if output_tokens is None:
        output_tokens = _number_field(value.get("output_tokens"))
    total_tokens = _number_field(value.get("total_tokens"))
    if total_tokens is None and (input_tokens is not None or output_tokens is not None):
        total_tokens = (input_tokens or 0) + (output_tokens or 0)
    result: dict[str, int | float] = {}
    if input_tokens is not None:
        result["inputTokens"] = input_tokens
    if output_tokens is not None:
        result["outputTokens"] = output_tokens
    if total_tokens is not None:
        result["totalTokens"] = total_tokens
    return result or None


def _number_field(value: object) -> int | float | None:
    if isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(value):
        return None
    return value


def _image_rows(result: Mapping[str, Any]) -> list[dict[str, Any]]:
    return [dict(item) for raw in as_list(result.get("images")) or [] if (item := as_mapping(raw)) is not None]


def _extracted_urls(result: Mapping[str, Any]) -> set[str]:
    return {
        str(item.get("imageUrl"))
        for item in _image_rows(result)
        if isinstance(item.get("text"), str) and item["text"].strip()
    }


def _has_image_text(result: Mapping[str, Any]) -> bool:
    return any(isinstance(item.get("text"), str) and item["text"].strip() for item in _image_rows(result))


def _merge_image_responses(primary: Mapping[str, Any], fallback: Mapping[str, Any]) -> dict[str, Any]:
    seen: set[str] = set()
    images: list[dict[str, Any]] = []
    for item in [*_image_rows(primary), *_image_rows(fallback)]:
        image_url, text = item.get("imageUrl"), item.get("text")
        if isinstance(image_url, str) and isinstance(text, str) and text.strip() and image_url not in seen:
            seen.add(image_url)
            images.append(item)
    result: dict[str, Any] = {
        "images": images,
        "rawText": "\n".join(
            value for value in (primary.get("rawText"), fallback.get("rawText")) if isinstance(value, str) and value
        ),
    }
    usage = _merge_usages([usage for candidate in (primary, fallback) if (usage := as_mapping(candidate.get("usage"))) is not None])
    if usage is not None:
        result["usage"] = usage
    return result


def _merge_usages(usages: list[Mapping[str, Any]]) -> dict[str, int | float] | None:
    totals: dict[str, int | float] = {}
    for usage in usages:
        for key in ("inputTokens", "outputTokens", "totalTokens"):
            number = _number_field(usage.get(key))
            if number is not None:
                totals[key] = totals.get(key, 0) + number
    return totals or None


def _response_error_suffix(response: httpx.Response) -> str:
    cleaned = " ".join(response.text.split())
    return f" - {cleaned[:500]}" if cleaned else ""


def _unsupported_temperature(message: str) -> bool:
    lowered = message.casefold()
    return "temperature" in lowered and ("unsupported" in lowered or "only the default" in lowered)


def _nullish(value: object | None, fallback: object | None) -> object | None:
    return fallback if value is None else value


def _text(value: object) -> str:
    return value if isinstance(value, str) else str(value or "")


AzureApiKeywordClassifier = AzureOpenAIProvider
